File: wiringPi/python/i2c_ads1115.py
#!/usr/bin/env python

'''
Name:    i2x_ads1115.py
Purpose: Read analog signal using DAC
         ADS1115 through i2c port
 Author:  Aldo Nunez
'''
# Import GPIO module
import wiringpi as wipi

def

def main ():
    # integer to float conversion constant
    VOLTS_PER_STEP =  6.144 / 32768.0

    '''
    ADS1115 configuration 
    OS   = 0x1       Single shot
    MUX  = 0x100     AINp = AIN0, AINn = GND
    PGA  = 0x000     Full Scale +-6.144 Volts
    MODE = 0x1       Power-down single shot
    DR   = 0x110     Data Rate 475 SPS
    COMP_MODE = 0x0  Traditional comparator with hysteresis
    COMP_POL  = 0x0  Active low
    COMP_LAT  = 0x0  Non-latching comparator
    COMP_QUE  = 0x0  Assert after one conversion
    '''
    config = 0x0
    config |= OS       | MUX2     | MUX1      | MUX0
    config |= PGA2     | PGA1     | PGA0      | MODE
    config |= DR2      | DR1      | DR0       | COMP_MODE
    config |= COMP_POL | COMP_LAT | COMP_QUE1 | COMP_QUE0

    # opens i2c-1 device
    int fd = wiringPiI2CSetup ( ADS_ADDRESS )
    if fd == -1:
        print ( stderr, "Couldn't open i2c-1 device\n" )
        return 1

    # writes config parameters
    if wiringPiI2CWriteReg16 ( fd, 1, config ):
        print ( stderr, "Error writing device\n" )
        return 1;

    # reads voltage as 16 bits integer
    input.data16 = wiringPiI2CReadReg16 ( fd, 4 )
    if input.data16 < 0:
        print ( stderr, "Error reading data\n" )
        return 1

    '''
    MSB in data8 [ 0 ] 
    LSB in data8 [ 1 ] 
    '''
    uint16_t int_val = 0x0;
    int_val |= input.data8 [ 0 ] << 8;
    int_val |= input.data8 [ 1 ];

    float_val = int_val * VOLTS_PER_STEP

    print ( "Input voltage: %4.3f Volts\n", float_val )

if __name__ == "__main__":
    main ()

File: wiringPi/c++/src/i2c_ads1115.cpp
//
// Program: i2c_ads1115.cpp
// Purpose: Read analog signal using DAC
//          ADS1115 through i2c port
// Author:  Aldo Nunez
//

#include<iostream>
#include<inttypes.h>
#include<unistd.h>
#include<iomanip>
#include<wiringPi.h>
#include<ads1115.h>

using namespace std;

int
main()
{
    // ADS1115 address 
    const unsigned int  ADS_ADDRESS = 0x48;

    // chooses a pin value > 63 
    const int pin = 64; 

    // integer to double conversion constant 
    const double VOLTS_PER_STEP = 6.144 / 32768;
    
    int16_t int_value = 0x0;
    double  voltage = 0.0;

    // sets up ADS1115 
    if ( !ads1115Setup ( pin, ADS_ADDRESS ) )
    {
        cerr <<  "Couldn't open i2c-1 device\n";
        return 1;
    }
    
    // configures gain control to 6.144 volts - 0 
    digitalWrite ( pin, 0 );
    
    // configures data rate to 475 SPS - 6
    digitalWrite ( pin + 1, 6 );

    while ( 1 )
    {
        // reads input 
        int_value = static_cast < uint16_t > ( analogRead ( pin ) );

        // converts to double 
        voltage = int_value * VOLTS_PER_STEP;
        
        // prints lecture 
        cout << fixed << setprecision ( 4 )  << \
        "Input voltage: " <<  voltage << endl;

        /* waits */
        sleep ( 1 );
    }

    return 0;
}

